src/jit/flowgraph.cpp

```diff
@@ -5746,14 +5746,36 @@ void          Compiler::fgFindBasicBlocks()
 
     if (compIsForInlining())
     {
+        bool hasReturnBlocks = false;
+        bool hasMoreThanOneReturnBlock = false;
+
+        for (BasicBlock* block = fgFirstBB; block; block = block->bbNext)
+        {
+            if (block->bbJumpKind == BBJ_RETURN)
+            {
+                if (hasReturnBlocks)
+                {
+                    hasMoreThanOneReturnBlock = true;
+                    break;
+                }
+
+                hasReturnBlocks = true;
+            }
+        }
+
+        if (!hasReturnBlocks)
+        {
+            impInlineInfo->inlineResult->NoteFatal(InlineObservation::CALLEE_DOES_NOT_RETURN);
+            return;
+        }
+
```

Comment:
As @jkotas already pointed out exceptions are very slow anyway. The typical throw sequence already needs at least 3 calls - new, ctor and throw - and the throw itself is very expensive.

Note that this does not affect methods like this one:
```C#
void ThrowIfNull(object value) {
    if (value == null)
        throw new ArgumentNullException();
}
```
Not inlining this would be bad since you'd pay the cost of a call even if `value` is not null.
```
```